hbase_storage/src/main/java/com/splicemachine/access/client/SkeletonClientSideRegionScanner.java

```diff
@@ -141,10 +143,44 @@ public boolean isFilterDone() throws IOException{
         return scanner.isFilterDone();
     }
 
+    private void fetchNextAndGiveBuffer(List<Cell> result) throws IOException {
+        assert result.isEmpty();
+        List<Cell> nextResult = new ArrayList<>();
+        boolean response = nextMerged(nextResult);
+        if (matchingFamily(nextResult, ClientRegionConstants.FLUSH)) {
+            // A flush should be returned before a potential partial result in the buffer
+            result.addAll(nextResult);
+        } else {
+            result.addAll(rowBuffer);
+            rowBuffer = nextResult;
+            if (!response) {
+                noMoreRecords = true;
+            }
+        }
+    }
+
     @Override
     public boolean nextRaw(List<Cell> result) throws IOException {
-        boolean res = nextMerged(result);
-        boolean returnValue = updateTopCell(res,result);
+        boolean firstCall = rowBuffer == null;
+        boolean response;
+        if (firstCall) {
+            rowBuffer = new ArrayList<>();
+            response = nextMerged(rowBuffer);
+            if (!response) {
+                result.addAll(rowBuffer);
+                rowBuffer = null;
+            } else {
+                fetchNextAndGiveBuffer(result);
+            }
```

Comment:
So if there are no more rows, we consume the rowBuffer and return, otherwise, we consume the rowBuffer and pre-fetch the next row into rowBuffer, right?

Reply:
yes